keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+  This feature is not yet implemented but per-Service enablement/disablement is
+  covered in depth as part of the test plan.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  The biggest risk here is that EndpointSlices may be created with hints some
+  but not all zones. This will be covered by kube-proxy falling back to all
+  endpoints if none have hints.
+
+* **What specific metrics should inform a rollback?**
+  If the proportion of `endpoint_slice_controller/syncs` with a "failure" result
+  is greater than 10%, a rollback may be considered. It is worth noting that
+  other issues can cause sync failures such as an out of date informer cache.
+  The key indicator should be a significantly elevated error rate when compared
+  with before the feature was enabled.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  This feature is not yet implemented but per-Service enablement/disablement is
+  covered in depth as part of the test plan.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features,
+  APIs, fields of API types, flags, etc.?**
+  Yes, this represents a replacement to the approach tracked with KEP 536. This
+  KEP included an alpha implementation but did not graduate beyond that.
+
+### Monitoring Requirements
+
+* **How can an operator determine if the feature is in use by workloads?**
+  If the `endpointslices_changed_per_sync` metric has a non-zero value for the
+  `auto` approach, this feature is in use.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to
+  determine the health of the service?**
+  - [x] Metrics
+    - Metric name: `endpoint_slice_controller/syncs`
+    - [Optional] Aggregation method: Counter
+    - Components exposing the metric: EndpointSlice Controller
+    - The relative failure rate over time can be used to track the health of
+      this controller.
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  As a starting point, it is likely reasonable for the EndpointSlice controller
+  to experience up to a 10% sync failure rate. This is largely related to it
+  trying to update stale EndpointSlices. When we are able to find a solution for
+  that issue the expected sync failure rate should be significantly lower. This
+  specific problem is most notable for large Services that have rapidly updating
+  endpoints.
+
+* **Are there any missing metrics that would be useful to have to improve
+  observability of this feature?**
+  None that I can think of.
+
+### Dependencies
+
+* **Does this feature depend on any specific services running in the cluster?**
+  No new dependencies.
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+  This may result in some additional calls to the EndpointSlice API, but expect
+  the increase to be minimal.
```

Comment:
Is endpoint-slice controller currently watching for nodes?
Is kube-proxy watching for its node?

If the answer for any of those is "no", then those are also new api calls (both fine, but should be stated here).

Reply:
Good point. [EndpointSlice controller already watches for nodes](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go#L78). Kube-Proxy does not by default but [does when ServiceTopology gate is enabled](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L754-L767). This would also be the same watch simply covering the node the instance is running on. Will update this to include that.

Reply:
It also has another watcher for its own node here

https://github.com/kubernetes/kubernetes/blob/97d40890d00acf721ecabb8c9a6fec3b3234b74b/cmd/kube-proxy/app/server_others.go#L386-L402

used if 
```
	if detectLocalMode == proxyconfigapi.LocalModeNodeCIDR {
		klog.Infof("Watching for node %s, awaiting podCIDR allocation", hostname)
		nodeInfo, err = waitForPodCIDR(client, hostname)
```

Reply:
This one seems to be short-living. But we should probably unify these.

Reply:
@robscott - please update :)

Reply:
Updated, thanks!